### PR TITLE
Add support for rust-analyzer

### LINKED ISSUE #2832: Support rust-analyzer
Add support for rust-analyzer

**Name:** rust-analyzer
**URL:** https://github.com/rust-analyzer/rust-analyzer

rust-analyzer is an experimental rust language server which is part of a rls 2.0 effort.

### PR DESCRIPTION
**Pull request by jonhoo**
Fixes #2832.

This is a very straightforward adaptation of the existing `rls` support.

### CODE REVIEW

File: ale_linters/rust/analyzer.vim [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+endfunction
+
+function! ale_linters#rust#analyzer#GetProjectRoot(buffer) abort
+    let l:cargo_file = ale#path#FindNearestFile(a:buffer, 'Cargo.toml')
+
+    return !empty(l:cargo_file) ? fnamemodify(l:cargo_file, ':h') : ''
+endfunction
+
+call ale#linter#Define('rust', {
+\   'name': 'rust-analyzer',
```
Comment (w0rp) [inline ale_linters/rust/analyzer.vim]: Let's name this `analyzer` so it matches the variable name and documentation sections.

File: doc/ale-rust.txt [hunk 2]
```diff
@@ -23,6 +23,11 @@ Integration Information
      over cargo. rls implements the Language Server Protocol for incremental
      compilation of Rust code, and can check Rust files while you type. `rls`
      requires Rust files to contained in Cargo projects.
+  3. analyzer -- If you have rust-analyzer installed, you might prefer using
```
Comment (cwndrws) [inline doc/ale-rust.txt]: This messes up numbering of linters. This should be 4 and `rustfmt` should be 5.

Comment (cwndrws) [inline doc/ale-rust.txt]: This seems to have already been messed up before you made your changes, but the line above this list says:
```
Note that there are three possible linters for Rust files:
```
but there are now 5 entries in the list and one of them isn't a linter, it's a fixer (rustfmt). The number should be changed to reflect the number of items in the list. These should either be split into two lists or the wording should be changed to reflect that there are both linters and fixers in here.

**Comment by jonhoo** [comment]
Thanks! Fix pushed.

**Comment by syphar** [comment]
@jonhoo thanks for doing this!

**Comment by kamek-pf** [comment]
@jonhoo I gave your branch a shot, RA seems to run as intended, but I can't get clippy to run.
I tried with `let b:ale_rust_analyzer_config = {'rust': {'clippy_preference': 'on'}}`
Maybe there's some sort of integration between rust analyzer and clippy to enable somewhere ?

**Comment by Arzte** [comment]
Heya is there any blockers on merging this?

**Comment by jonhoo** [comment]
Not on my part. I no longer use ale, but last I checked the patch worked fine :)
I _think_ that this line may need to be updated if the server binary has gotten a new name (which I remember there being some discussion of):
https://github.com/dense-analysis/ale/blob/715733f44def5e162d90184ae12886c54adeeb5b/ale_linters/rust/analyzer.vim#L4

It'd be good if someone actively using this could try it out and verify :+1:

**Comment by Arzte** [comment]
Applied this patch to my local copy of ale and set it to use rust-analyzer, seems to be working perfectly for me. 👍

**Comment by Shadow53** [comment]
The upstream binary name for rust-analyzer *has* been changed to rust-analyzer. That may be why it didn't work for Arzte.

**Comment by jonhoo** [comment]
@Shadow53 thanks! I've pushed a commit that changes the default.

File: test/command_callback/test_rust_analyzer_callbacks.vader [hunk 3]
```diff
@@ -0,0 +1,20 @@
+Before:
+  call ale#assert#SetUpLinterTest('rust', 'analyzer')
+
+After:
+  call ale#assert#TearDownLinterTest()
+
+Execute(The default executable path should be correct):
+  AssertLinter 'ra_lsp_server', ale#Escape('ra_lsp_server')
```
Comment (Arzte) [inline test/command_callback/test_rust_analyzer_callbacks.vader:8]: This needs to be changed to ``rust-analyzer to fix the tests

**Comment by 1wilkens** [comment]
Anything blocking this? I would really like to use `rust-analyzer` "out-of-the-box" with ale :)

**Comment by jonhoo** [comment]
@1wilkens not as far as I am aware — this is ready to review/merge from my side.

**Comment by RyanSquared** [comment]
Cheers! :beers:

**Pull request closed without merging**